src/pepper/parser.py

```diff
@@ -182,35 +226,94 @@ def p_expressions(p):
     p[0].append(p[2])
 
 
-def p_whitespace(p):
+def p_identifier_call(p):
     """
-    code_expression : WHITESPACE
+    safe_code_expression : IDENTIFIER code_expression_parenthetical
     """
-    p[0] = ast.WhiteSpaceNode(p[1])
+    print(f"macro call with ident {p[1]} and args {p[2]}")
+    p[0] = ast.IdentifierNode([p[1]], args=p[2])
+
+
+def p_code_expression_to_safe(p):
+    """
+    code_expression : safe_code_expression
+    """
+    p[0] = p[1]
 
 
 def p_statement_to_identifier(p):
     """
-    code_expression : IDENTIFIER
+    safe_code_expression : IDENTIFIER
     """
     p[0] = ast.IdentifierNode([p[1]])
 
 
+def p_expression_to_list_of_something(p):
+    """
+    code_expression_parenthetical : '(' list_of_expressions ')'
+    """
+    p[0] = p[2]
+
+
+def p_whitespace_unsafe(p):
+    """
+    safe_code_expression : WHITESPACE
+    """
+    p[0] = ast.WhiteSpaceNode([p[1]])
+
+
 def p_expression_to_string_lit(p):
     """
-    code_expression : STRING_LITERAL
+    safe_code_expression : STRING_LITERAL
     """
     p[0] = ast.StringLiteralNode([p[1]])
 
 
-def p_statement_to_ascii_literal(p):
+
+def p_expression_list_singleton(p):
     """
-    code_expression : '<'
+    list_of_expressions : safe_code_expressions
+    """
+    p[0] = [ast.LinesNode(p[1])]
+
+
+def p_expression_list_empty(p):
+    """
+    list_of_expressions :
+    """
+    p[0] = []
+
+
+def p_expression_list_multiple(p):
+    """
+    list_of_expressions : list_of_expressions ',' maybe_space safe_code_expressions
+    """
+    p[0] = p[1]
+    p[0].append(ast.LinesNode(p[4]))
+
+
+# don't  mind me, just duplicating code...ugh
+def p_safe_expressions_empty(p):
```

Comment:
I commented this before if i'm not sure it got lost within the additional pushes, but what makes safe_expression different from code_expression?